alog2.go

```diff
@@ -0,0 +1,123 @@
+package algo
+
+import (
+    "fmt"
+    "bytes"
+)
+
+type Node struct {
+    name string
+    children []Node
+}
+
+//Holds elements of Stack.
+type StackElement struct {
+    nodes []*Node
+}
+
+func (self *StackElement) pop() *Node {
+    size := len(self.nodes)
+    if (size > 0) {
+        size--
+        result := self.nodes[size]
+        self.nodes = self.nodes[:size]
+        return result
+    }
+    return nil
+}
+
+//Representation of stack.
+//Each level corresponds to Node's nesting level
+//where number of nodes can reside.
+type Stack struct {
+    inner []StackElement
+}
+
+func (stack *Stack) last() *StackElement {
+    return &stack.inner[len(stack.inner)-1]
+}
+
+func (stack *Stack) len() int {
+    return len(stack.inner)
+}
+
```

Comment:
And if you had used only one structure - you wouldn't have needed to implement all these methods.